### ISSUE TITLE
$ are not allowed in filename of a filegroup

### ISSUE THREAD

**Issue by damienmg** [google/bazel#374]
Ultimately any character can be part of a filename. We should probably allow that.

Some mangling to generate the corresponding label should probably be done.

Original report on the mailing-list:
 https://groups.google.com/d/msgid/bazel-discuss/CAN0GiO3__5jXo5rZqroSj0mFxpqCzUZZVkY%3DSNsJK1%2BZ1BdJLg%40mail.gmail.com

**Comment by abergmeier** [comment]
1. So are we talking Unicode?
2. Where does _any character_ stop?
3. how do you treat characters that are not allowed on certain platforms?
4. When using mangling, how do you handle collisions?

**Comment by kayasoze** [comment]
In POSIX, filenames are "bags of bytes"--there is no encoding; however, `NUL` and `/` are not allowed.  Windows has a few more restrictions.  Perhaps the BUILD file should be parsed in the encoding of the system locale, usually `UTF-8`, and filenames run though a `ValidForCurrentPlatform()` function which checks for disallowed characters.  However, opting for strict platform neutrality in this way means that Bazel would have to represent filenames as a bag of bytes and not a Unicode string, as there is no guarantee that the filename will roundtrip through Unicode correctly.  The problem can probably be simplified by restricting filenames to be `UTF-8 or `UTF-16`, which should cover most people's needs even though that's not strict POSIX.

**Comment by ulfjack** [comment]
Well, I think we can probably require valid UTF-8 file names and strongly recommend that people use UTF-8 for their file system. For labels / BUILD files, we probably need an escaping scheme, at least for the control characters. If there's a file that isn't valid UTF-8, we give an error message?

**Comment by btelles** [comment]
Our company codes mainly in C++, but our frontend uses a lot of JS and nodejs modules which have all sorts of characters in the filenames--for example, -, #, @, (, and ).

Right now this is a major blocker for getting all our codebase under one build system since we can't reference files with semi-special characters. I don't think Bazel should decide what characters are acceptable in file names, as that reduces file names to those that fit both (1) supported languages and (2) supported platforms. This seems unnecessarily restrictive, and is becoming a major pain point for us.

**Comment by ulfjack** [comment]
Agreed. Unfortunately, it's a bit tricky to fix, as a lot of code assumes that the mapping from labels to file names (and vice versa) is trivial, and doesn't require escaping. Any suggestions on an escaping scheme?

**Comment by damienmg** [comment]
URL based?

**Comment by abergmeier** [comment]
You mean an own URI scheme? Sounds good.

**Comment by damienmg** [comment]
I mean replacing special characters by %XX where XX is the UTF-8 code in hexa.

**Comment by ulfjack** [comment]
Sorry, I won't be able to work on this. @philwo had an interest, maybe he can make some progress here. :-/

**Comment by kayasoze** [comment]
This blocks our Bazel deployment as well.

**Comment by RonnieAtOracle** [comment]
This is blocking us.  We have a templating system where we need to build our template files.  The filenames themselves contains template variables (e.g. ${ServiceName}.java ).   Both $, {, and } are not supported by Bazel in file names.

**Comment by philwo** [comment]
I totally agree that this is important, should be done, I want this myself, however I don't have the time to work on it in the coming months, thus I have to unassign it.

**Comment by DemiMarie** [comment]
Here is my proposal:

- Metacharacters (`:`, `%`, `=`, any others) must be %-encoded
- All other characters are allowed.  This includes Unicode characters.
- Non-UTF8 names are not allowed, even if escaped.  This is because there is no good way to handle them cross-platform, nor to display them to the user.

**Comment by mihnita** [comment]
Plain ASCII (and even that partial) makes this feels like we are in the early 90s.

There are reasonable ways to handle that.
For POSIX using the default locale would be good enough.

If my project is C/C++, and it is cross-platform, and I have problems handling Unicode, then I will not use Unicode in file names. And the fact that bazel "explodes" is not such a problem.
But if I use something like Java, then "it just works", and bazel would work too.

Even better would be to to allow for a character-set option in the project file.
This is what maven does. And what Java does with -Dfile.encoding=UTF-8
So if it is there, use it. If not, then take the system charset.

I did not move one project to bazel because test units check that Unicode file names work.
So the files are there, make it through git, work with maven and ant and gradle and java.
But bazel fails because there is an "@" in the file name... Which is supported on all OSes.

**Comment by davido** [comment]
Missing support for "$" is really painful, as this char is a valid one in the Java file name, yet we are getting this:

```
ERROR: /home/davido/projects/prolog-cafe/BUILD:51:1: //:builtin: invalid label 'java/com/googlecode/prolog_cafe/builtin/PRED_$atom_type0_2.java' in element 0 of attribute 'srcs' in 'java_library' rule: invalid target name 'java/com/googlecode/prolog_cafe/builtin/PRED_$atom_type0_2.java': target names may not contain '$'.
```

Buck build tool supports this out of the box and it really sucks, to go the srcjar zip route in Bazel.

**Comment by damienmg** [comment]
FTR this has become high priority and is now my main tasks. But this might take a long time before it is default (because we have to cover all corner cases).

**Comment by davido** [comment]
//CC @dborowitz @spearce

Given this major drawback of current Bazel implementation, I'm procrastinating to migrate the last Gerrit Code Review sub-project from Buck to Bazel: `prolog-cafe`, and wait with this migration, until it is fixed. Thoughts?

**Comment by dborowitz** [comment]
When we use Prolog Cafe from blaze, we get around this limitation by using a Python script to build a srcjar (zip file) containing all the source files outputted by the Prolog Cafe compiler. This srcjar is a valid input to a java_library rule. I can share this Python script downstream in our Prolog Cafe fork.

Not directly related to this bug, but another issue we ran into with blaze is it doesn't like having multiple top-level classes in a single .java file; I forget exactly what happens, but it either fails fast or messes with caching. The same Python script that creates the srcjar also takes care of splitting up top-level classes into their own .java files.

**Comment by damienmg** [comment]
Update: https://docs.google.com/document/d/1ducs753wqYoE6ibxYYwnVlX7RSw7YQiVRNiMz-hgYQE/ is the draft design doc. The quoting is subject to change but the rest should be pretty stable.

**Comment by alexeagle** [comment]
This also affects Angular / TypeScript using Bazel.

We install packages from NPM to run in nodejs. NPM uses the @ character as a "scoped package", eg. `@types`. When installing external dependencies with my repository_rule, the packages are installed, and I write a BUILD file into `node_modules` to allow depending on the files:
```
project/
  node_modules/
    BUILD
    @types/
      foo/index.d.ts
```

However, I cannot make a label in this location that points to `@types/foo/index.d.ts` because it gets interpreted as a workspace name: `invalid label '@types/foo/index.d.ts' in element 5 of attribute 'srcs' in 'filegroup' rule: invalid repository name '@types/foo/index.d.ts': workspace names may contain only A-Z, a-z, 0-9, '-', '_' and '.'.`

Workaround for me is to put the BUILD file one level higher, but this is really worse because that location should be controlled by the user, not my repository rule.

**Comment by damienmg** [comment]
'@' is not allowed in the package name, it can probably be added immediately (only you need to add a leading colon to avoid confusion for remote repository)

**Comment by alexeagle** [comment]
@damienmg is that something we could get in now? It would be really helpful, my workaround is not that clean. If it's useful and you point me to the right spot I can make the contribution.

**Comment by davido** [comment]
Almost two years later after this bug was created, any progress here?

As a compromise, the most annoying limitation is not supporting the "$" character for Java eco-system, as this is a valid character in class names. My suggestion would be to forget all other characters but hard code the support for "$" in target names in Bazel, and deliver that feature in the next 2 months and not **all** chars in the next 20 years.

Two overcome this limitation we are forced to "hide" the file names from Bazel now: [1].

Buck version:

```python
java_library(
  name = 'builtin',
  srcs = glob([SRC + 'builtin/*.java'], exclude = REPL + IO) +
  [
    ':builtin_srcs',
    ':system_srcs',
  ],
  deps = [':lang'],
)
```

Bazel version:

```python
# TODO(davido): Fix that mess when this major Bazel bug is fixed:
# https://github.com/bazelbuild/bazel/issues/374
# That why I left the original :builtin rule from Buck, so that
# you can feel my pain, to emulate the glob with zip, to hide
# the files that contain '$' from Bazel.
genrule(
    name = "builtin_srcjar",
    outs = ["builtin.srcjar"],
    cmd = " && ".join([
        "TMP=$$(mktemp -d || mktemp -d -t bazel-tmp)",
        "ROOT=$$PWD",
        "cd java",
        "zip -q $$ROOT/$@ com/googlecode/prolog_cafe/builtin/*.java",
        "zip -qd $$ROOT/$@ com/googlecode/prolog_cafe/builtin/PRED_\$$write_toString_2.java %s" % " ".join([s[5:] for s in IO]),
    ]),
    local = 1,
)

java_library(
    name = "builtin",
    srcs = [
        ":builtin_srcjar",
        ":builtin_srcs",
        ":system_srcs",
    ],
    deps = [":lang"],
)
```

What a mess?!

* [1] https://gerrit.googlesource.com/prolog-cafe/+/3d340c5a26080d679bf5e2baf90b0c8a0372f025/BUILD#57

**Comment by alexeagle** [comment]
@damienmg any update?

**Comment by damienmg** [comment]
Sorry I haven't got time to work on this recently, I have been on holiday and working from China right now. I need to do further investigation to allow `@`.

Note that ` `, `$`, `(` and `)` are now allowed in globs and filegroup though you might hit other problem along the way, it works fine for creating jar that takes those as argument.

We still have to consolidate the design to be able to move forward with allowing any characters in filename.

**Comment by davido** [comment]
> Note that , $, ( and ) are now allowed in globs and filegroup though

Thanks. It worked here.

**Comment by macrael** [comment]
Has this change been released yet? I don't like relying on a custom built bazel if possible.

**Comment by damienmg** [comment]
Yes

**Comment by davido** [comment]
> Yes

@damienmg Can you mention in what Bazel release support for `$` character was actually added? I was using 0.5.2rc2 and it worked here (https://github.com/GerritCodeReview/prolog-cafe/commit/9eaffdc036c405d0199d3e2a441896541b6e7c1c).

**Comment by damienmg** [comment]
I believe it is since 0.4.5 but you'll have to check the release notes for
which one exactly.

**Comment by ddwolf** [comment]
My company is pushing R&D teams to use bazel. We have a lot of projects with files have names contain non-ascii charactors, so supporting non-ascii names is very important for us. We have a workaround:remove the restriction and transfer filenames into `locale` format, hope these feature coud be supported in a short future

**Comment by damienmg** [comment]
Can you send a PR to discuss the possibility of accepting it? As far as I
understand our only problem with those characters is actually printing
them, there is no reason not to allow them. Add @philwo and I to the review.

**Comment by philwo** [comment]
@ddwolf I agree with Damien - it would be great if you could share your workaround as a pull-request, just to see what kind of changes are required to make it work for you.

I think we're currently not dealing with character encoding in a well defined way in Bazel - some code knows about the "latin-1 tunnel" to deal with unknown characters, other code treats everything as UTF-8, ... but I'm happy to audit it and see how we can clean it up.

**Comment by ddwolf** [comment]
@damienmg @philwo thank you for you replies, Jianfeng Yu would issue a pull request shortly after. have you considered to import escape charactors to make "label name" support any charactors?

**Comment by damienmg** [comment]
Yes this is the long term solution, but it takes time to agree on how
exactly to do it (and there are always other fire to fight).

**Comment by damienmg** [comment]
Ok we just had a discussion with @philwo in which we agreed on how to move forward (Philip, correct me if I missed something):

- We are going to reserve '%' for quoting of labels. '%%' is going to be used for a percent in the label and '%:' is going to be used for colons in label. All other character will stay unquoted for now. We might add more quoting possibilities new use case requires it.
- We are going to allow all characters in filename, including non ASCII, right now. And add test cases we already had in [this doc](https://docs.google.com/document/d/1MF-UVcTa7CvalZ-bFr1nRyGwHIfJz9U7CaMjjE72YQ0) plus any other test case we encounter. All tests will be extendable with characters / string to inject in the label to make sure we can add any filenames.
- We are going to fix any encoding issue we encounter when we had more test case. We want to use the local to read the filename from the file system. Maybe we will need to add a special method to delete files that have filenames not allowed with the locale encoding. We might want to mess around with locale in our test cases. [1]

[1] If we have a filename with weird characters, we need to write it with shell quoting in our tests (e.g. \xF0\x0B\xA1\x2), so maybe we don't need to mess around with locale.

**Comment by damienmg** [comment]
Also we are going to track the encoding on this bug too so adding philwo as assignee because he also wants to do some work to support it.

**Comment by siddharthab** [comment]
Joining the group here. We are affected by the % character being disallowed. We are using an external http archive, so can not change anything in the files or create zips of files.

**Comment by kamalmarhubi** [comment]
Just found this bug. @damienmg could you update the milestone on this with a current estimate? Thanks!

**Comment by damienmg** [comment]
Milestones are no longer relevant. Nobody's actually working on that at the moment AFAICT

**Comment by fiorix** [comment]
Landing here from #167.

Trying bazel for the first time today. My code lives in `~/Dropbox (Personal)/foobar` and bazel refuses to build it because the workspace's path has a space on it. I tried creating a symlink to `~/foobar` but didn't help. All I get is this:

```
ERROR: bazel does not currently work properly from paths containing spaces (com.google.devtools.build.lib.runtime.BlazeWorkspace@24eea938).
```

This is tricky because neither bazel supports space in paths and dropbox won't let me rename that directory. :)

**Comment by appsforartists** [comment]
@fiorix I'm painted into the same corner.  I need to develop an Adobe plug-in inside `~/Library/Application Support`, but Bazel chokes on the space.  I'm going to poke at the Bazel source and see if I can find an easy way to suppress that error.

**Comment by appsforartists** [comment]
Looks like the no spaces error was introduced in #167 by @kchodorow 3 years ago.  Based on the more recent comments on this issue, I'm presuming that this error is obsolete and can now safely be removed.

I don't have my machine set up to build Bazel now, but I might take a stab at removing this error and seeing what breaks.  😃

**Comment by ulfjack** [comment]
There are a couple things in Bazel that may not be whitespace-safe.

Bazel generates runfiles and fileset manifests in a text format which uses whitespace as a separator. There are also a few generated wrapper scripts, especially for java_binary and py_binary, as well as a test wrapper script (test-setup.sh), which may not be whitespace-safe. I don't expect any problems with the Java code as such.

**Comment by ulfjack** [comment]
I started a change that allows more characters in labels, but I think it still excludes whitespace - though note that that is separate from whitespace in the path to the workspace itself.

**Comment by appsforartists** [comment]
Thanks @ulfjack.  A few things I'm not clear on:

- Are you still actively working on this?  Seems that this bug has passed through a handful of people (Kristina, Damien) who no longer work on Bazel; it may need an owner.
- Do you expect it's currently safe to use spaces in paths?
- What would need to change to get an LGTM removing the assertion that a path doesn't contain spaces?

**Comment by ulfjack** [comment]
I am not currently working on this.

By all means, give it a try. If we audit (and, if necessary, fix) the places mentioned above, and maybe add an integration test, and it works, that's good enough for me. There may be hidden bugs, but we won't find them if nobody is trying it.

**Comment by kellycampbell** [comment]
I ran into this problem and it was difficult to debug. 

```
ERROR: /project/tools/protobuf/BUILD:11:1: Creating runfiles tree bazel-out/k8-fastbuild/bin/tools/protobuf/pbts.runfiles failed: Process exited with status 1: Process exited with status 1
Target //tools/protobuf:pbts failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

`--verbose_failures` gave no information. No other options I could find gave any extra output about the error. I finally determined that it was a new node_modules dependency that had been added and then within the runfiles MANIFEST.tmp file, the last line was a file with spaces in the filename. If I removed the files with spaces, then the problem was corrected.

Is there some other debug or verbose flag that would give better information about this failure?

**Comment by siddharthab** [comment]
I think 0.16 gives a better error message. Just ran into the same error message a few days ago, and colleague received a proper reason by updating bazel on his machine. @skinner

**Comment by mfarrugi** [comment]
Echo'ing @kellycampbell, not easy to debug that bazel is unhappy with whitespace's in python's setuptools.

Luckily those files are safe to ignore..
```python
py_runtime(
    name = "python3",
    files = glob(
        ["dist/**"],
        exclude = [
            # These bad boys have spaces in their name, and this kills the bazel.
            # https://github.com/bazelbuild/bazel/issues/374
            # https://github.com/pypa/setuptools/issues/746
            "dist/lib/python3.6/site-packages/setuptools/command/launcher manifest.xml",
            "dist/lib/python3.6/site-packages/setuptools/script (dev).tmpl",
        ],
    ),
    interpreter = "dist/bin/python3",
    visibility = ["//visibility:public"],
)
```

**Comment by meisterT** [comment]
Is this still a problem with Bazel 0.17.2?

**Comment by adam-azarchs** [comment]
Asking for "any" character is perhaps making this feature request more complex to implement than is really required for most users.

I think it would be reasonable, from a perspective of maintaining the cross-platform objectives of bazel, to simply allow the full set of ascii characters which are allowed in both Windows and Posix file names (that is, everything except `/`, `\`, `?`, `%`, `*`, `:`, `|`, `"`, `<`, `>`).  That would make it much easier to integrate with third-party code which uses such characters in file names.  In particular `+`, `(` and `)` are common everywhere, and npm frequently uses `@`.  Spaces are a bit tricky because they need to be escaped correctly when building command lines, perhaps, but nothing about it seems insurmountable.

**Comment by mihnita** [comment]
_"full set of ascii characters which are allowed in both Windows and Posix file names"_

Why _"full set of ascii characters ... allowed"_ and not _"full set of characters ... allowed"_ ?
Why ASCII only? Windows and POSIX allow for Unicode characters, for a very-very long time.

Since we need a way to escape spaces, and figure out all the places we need to patch to make it work, then we can just escape Unicode. It's not much more work. It is not the escaping that is hard, it is 

If this would be a tool that is someone's toy project, whatever.
But if the intention is to be adopted and used, then let's make it be the best that it can be!

It would be a shame to keep it ASCII, this is (almost) 2019, not 1990s :-)

**Comment by steeve** [comment]
Note that this will break rules_go when Go 1.12 is out due to the presence of tests with utf8 chars in the file:
https://github.com/golang/go/tree/master/test/fixedbugs/issue27836.dir

**Comment by aiuto** [comment]
@philwo Is this really your problem, or do you want to send it to others?

**Comment by philwo** [comment]
@aiuto Thanks for checking in. Yep, I'm the owner of this and fine with it. I also have ideas how to fix and test this, but so far other issues were more pressing. :|

**Comment by jin** [comment]
cc @alandonovan

**Comment by mzeren-vmw** [comment]
Somewhat related: C++ SG15 has been discussing how to name files in JSON, and current thinking is use JSON UTF-8 for "bags of bytes" that roundtrip through UTF-8 (with no normalization allowed) and arrays of integers for everything else. https://wg21.link/P1689.

**Comment by alandonovan** [comment]
I have a change to Blaze (Googlers: cr/263443121) that makes it use UTF-8 for all external text strings, and UTF-16 for all internal strings. If all goes well it should land this week.

**Comment by dev-guy** [comment]
This means we still can't use Bazel to build boost 1.70.0 per #8108

**Comment by jsharpe** [comment]
This also fails for trillinos: @trilinos//:all: invalid label 'packages/kokkos/core/src/eti/ROCm/Kokkos_Experimental::ROCm_ViewCopyETIInst_int64_t_double_LayoutLeft_Rank1.cpp' in element 14026 of attribute 'srcs' in 'filegroup' rule: invalid target name 'packages/kokkos/core/src/eti/ROCm/Kokkos_Experimental::ROCm_ViewCopyETIInst_int64_t_double_LayoutLeft_Rank1.cpp': target names may not contain ':'

**Comment by thomaslangston-mongodb** [comment]
Just adding another example of an issue this causes.

I'm trying to see if I can download Chromium as an http_archive for a Bazel target for our end to end tests. However, at least on mac, the Chromium app package contains directories and files with spaces in the name (e.g. 'Chromium Framework.framework').

I could create a glob that specifies down to the correct folder, but then if I update chromium versions and the folder structure changes, my target will break. I haven't been able to find a work around yet.

**Comment by Icantjuddle** [comment]
Would like to offer example of an issue this causes:

I'm trying to apply a `filegroup` to the output created the Unity game engine. They use some files with spaces whose names are hardcoded into the engine. I cannot change them without breaking the engine. I have a hack to get around this for now but it is unfortunate that bazel is incapable of tracking an essential file.

What I ended up doing is adding an Unity editor plugin to rename copy the file(s) to the same place with spaces replaced by underscores after building. Then, before running I have a script in bazel symlink them back to their original names so it can be found by the game engine. I'm sure there is a better way, but if a file could have spaces it would mean I wouldn't need any hacks.

**Comment by brandjon** [comment]
Update: This issue was mentioned in connection with Bazel's handling of unicode strings (e.g. #12986), but is separate in that it asks for an extension of the label syntax.

This is a pretty significant design question, seeing as how fundamental labels are to Bazel. A URL-like percent encoding scheme (with normalization toward avoiding unnecessary escapes, and capitalizing hex digits) has the merit of simplicity and minimalism. Whereas allowing all but an excluded set of characters would allow for a more natural representation of non-latin filenames. Either way would deserve a lot more consideration than I think we're able to give to it at the moment, hence the P4 label.

In contrast, improving our unicode string handling is on the schedule for this year.

**Comment by brandjon** [comment]
To clarify: I interpret this issue as requesting a label syntax that admits filenames with unicode data. The broadest view is that arbitrary binary data should be allowed, to fully support Unix where everything but `NUL` (U+0000) and `/` (U+002F) is allowed.

Although it's currently not the case, we intend to fix Bazel to use UTF-8 encoding for BUILD/.bzl files (and probably miscellaneous other I/O boundaries, like the command line). Meanwhile, the Starlark specification is being clarified to add a `bytes` data type and better specify unicode string behavior. Note that for Bazel and the Java Starlark interpreter, string values may hold arbitrary binary data (in 16-bit form, just like Java strings), although string literals in BUILD/.bzl files are restricted to unicode data (no surrogate code points or values beyond U+10FFFF).

All this is to say that we retain the capability to read/write arbitrary binary data as strings in Bazel (and in addition, Starlark can do so in a portable way using the bytes datatype). I don't think there's anything requiring us to enforce UTF-8 for file names, or any reason to impose one platform's filename restrictions on another.

**Comment by janakdr** [comment]
Just to clarify, would a successful resolution of this issue include filenames that have colons in them? I expect yes, but not positive.

**Comment by adam-azarchs** [comment]
It would have to, in order to support my use cases which include e.g. perl files which frequently contain `::`.

**Comment by haxorz** [comment]
Re colons - What would the plan be for existing code, both inside and outside of Bazel, that splits a label string on the color char to separate out the package name and target name? Presumably there will be a Bazel flag like `--incompatible_allow_colon_chars_in_target_names`, but it seems difficult to be able to flip that flag's value.

**Comment by aiuto** [comment]
If we were going to do this, I would approach it by requiring that Bazel usage escape the : which is not the delimiter. So  `//foo\:bar:baz\:1`, is the target `baz:1` in the package `foo:bar`
We could probably make this change to Bazel. Maybe even very cleanly in a few places. And then the real fun begins.

I have no way to estimate the number of scripts and rules that could break if users actually took advantage of this. OTOH, it is probably not our problem. That is, if a user wants colons in file names and they write their own scripts that fail, it is their own problem.  

We'll probably also have issues around the meaning of \ on the command line.  Because of shell escaping, you will probably have to type   `//foo\\:bar:baze\\:1.`  That's easy and obvious, but I am sure it will generate support questions forever, so it has a little extra hidden cost.

**Comment by adam-azarchs** [comment]
That escaping is also an issue in starlark:
```starlark
filegroup(
    name = "perlio_mmap",
    srcs = [
        "PerlIO\:\:mmap.3",
    ],
)
```
would be a syntax error - you'd need to either make it an `r""` string in the BUILD file or use `\\` there as well.  The problem with this approach is that it gets more confusing if you for some reason had a filename with literal `\:` in it.  Now it would be `"foo\\\\\\:thing"`, or `r"foo\\\:thing"`?  I think, if I counted up my slashes correctly?  Which makes my point, I think.

An alternative approach might be to offer a wrapper function that makes a string be treated as a file rather than a label, e.g.
```starlark
filegroup(
    name = "perlio_mmap",
    srcs = [
        file("PerlIO::mmap.3"),
    ],
)
```
Which could just internally be e.g.
```starlark
def file(s):
    s.replace(":", r"\:")
```
but could also maybe be implemented by having bazel bypass whatever logic is used to try to guess whether a string refers to a label or a file, which seems to me at least to be the cleaner solution, though probably much harder to actually implement.

**Comment by adam-azarchs** [comment]
I'm reconsidering what I said earlier about allowing `:`.  Yes, it's in common usage in filenames in perl code, however given bazel is intended for cross-platform support, there's also an argument against supporting filenames which wouldn't be [legal in Windows](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions), specifically those containing any of `<>:"|?*\`.  Most of those are also problematic in cases of careless rule authors writing shell scripts.  Though there's a limit to how much bazel can protect them there[^1], at least if `"` isn't legal in target names then it's always safe to `"`-quote them.

[^1]: I'd really rather not to also prohibit `';&#()[]{}` or whitespace, all of which are also potentially problematic for a carelessly written shell command. I think whitespace is probably the most important use case for loosening the current restrictions, as there are a lot of use cases where file names become user-visible and normal humans don't automatically translate `_` into spaces.

**Comment by mihnita** [comment]
> given bazel is intended for cross-platform support, there's also an argument against supporting filenames which wouldn't be [legal in Windows](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions),

**TLDR: I think it is not the job of the build tool to force naming conventions on a project**

A project might be Linux and Mac only, no Windows.
Or might be Linux only, but using bazel just because they like it.
Or because use bazel because they us a mixture of targets / programming languages.
Or just because they know it (used it in other projects). Or want to learn.

The file naming rules would be selected by the developers looking at the big picture:
what they need to test, tools they use, version control tools, etc.
If each tool I use tries to force on me its own opinions on what's "the right thing" I end up with [A-Z0-9_]

Should a tool disallow the existence of the same file with different case?
(because there are case insensitive file systems)
If yes, what version of the case mapping table? (it can be system dependent)

Should a tool force a certain Unicode normalization?
Because the MacOS file system normalizes file names to decompose format.
(and what version of the normalization table?)

What about a Mac OS accessing a disk shared by a Linux machine using Samba (a Windows protocol)?

Long story short: not a bazel responsibility to force any limitations on what I can do or not.
IF bazel wants to help devs do "the right thing" then it can be as a validation step that can be disabled.
Not as a design limitation frozen in code.

**Comment by thomaslangston-mongodb** [comment]
>  Ideally any name would be supported, but if certain patterns are difficult to support, if they would anyway be illegal on some platforms that might be a sufficient excuse to not support them.

Illegal on some platforms is not a sufficient excuse to not support a filename pattern. Some filenames are system provided and cannot change to match the build system's preferences. Choosing some subset of valid filename patterns is equivalent to not having full support for any platform that is not encapsulated by the entire subset. We should be supporting the superset of all filename patterns instead.

**Comment by atobiszei** [comment]
Encountered this issue while using aws-sdk-cpp as a dependency. Luckily in that case it was only test file that we could safely remove with `patch_cmds` parameter in `git_repository`. Leaving it for anybody trying to use aws-sdk-cpp with bazel. It would be nice to see it resolved.

**Comment by snakethatlovesstaticlibs** [comment]
We're running into this in erlang 26 which bundles emoji filenames: https://github.com/erlang/otp/blob/master/lib/stdlib/test/edlin_expand_SUITE_data/.hidden%F0%9F%98%80_file

**Comment by fmeum** [comment]
@atobiszei @snakethatlovesstaticlibs Both your cases sound like you are in a slightly different situation: Files with complex file names are part of some external dependency you consume, but aren't relevant to its build (for example, because they are part of the test suite). The errors arise while extracting an archive. This use case should be fully supported in Bazel 6.4.0 and higher, could you give that a try?

**Comment by snakethatlovesstaticlibs** [comment]
@fmeum I worked around this by adding the test directories to an `exclude` in a `filegroup`, and it seems to be working now (because as you mentioned, the files are part of tests, so I don't need them)

For reference, the error I saw was:
```
ERROR: .../bazel_build/external/erlang/BUILD.bazel:10:15: Foreign Cc - Configure: Building erlang failed: error reading file '@@erlang//:erts/test/erlc_SUITE_data/src/😀/erl_test_unicode.erl': .../bazel_build/external/erlang/erts/test/erlc_SUITE_data/src/😀/erl_test_unicode.erl (No such file or directory)
```

**Comment by fmeum** [comment]
After a series of patches, Bazel 8 should now support the following on Linux, macOS and Windows (Windows build 1903 and higher):

* all Unicode characters except control characters, `:`, `\` and, in some positions, `.`, in file names that are also labels (i.e., can be referenced in BUILD files)
* all Unicode characters in other file names
* all Unicode characters in runfiles (including spaces and newlines)
* all Unicode characters in the output base or workspace path (including spaces)

You can try this with Bazelisk by setting `USE_BAZEL_VERSION` to the hash of the latest commit [on the release branch](https://github.com/bazelbuild/bazel/tree/release-8.0.0).

Please file a separate issue if you run into any problems with the above.

**Comment by phst** [comment]
> After a series of patches, Bazel 8 should now support the following on Linux, macOS and Windows (Windows build 1903 and higher):

Thanks a lot! Is there some documentation about this? Specifically:
Assume I have a file whose name appears as `☃.txt` in a graphical file explorer (Windows Explorer on Windows, Finder on macOS, Nautilus or similar on GNU/Linux systems). This filename will be encoded as UTF-16 on Windows, UTF-8 on macOS (actually UTF-16 NFD on HFS+ AFAIK, but the macOS POSIX API will present it as UTF-8 NFD), and (typically) UTF-8 on GNU/Linux (assuming LC_CTYPE is set to something with UTF-8).

- In a BUILD file, how do I reference such a file in a cross-platform way? Is it sufficient to encode the BUILD file as UTF-8, despite https://bazel.build/versions/8.0.0/concepts/build-files saying the BUILD files are interpreted as Latin-1?
- In a Starlark file, how does that filename appear in a string? What are the string's elements? (My assumption is that the string's elements are the UTF-8 code units as stored in the BUILD file.)
- How will that filename be encoded in runfiles manifests? As UTF-8, even on Windows? Is there any difference between host and exec configs?

**Comment by fmeum** [comment]
The current situation is that all Starlark and other Bazel files (e.g. `.bazelrc`) are assumed (but not enforced) to be encoded in UTF-8. File system paths are reencoded into the appropriate OS/FS-dependent encoding when Bazel interfaces with the OS/FS. 

Starlark strings behave as UTF-8 byte arrays, with no notion of Unicode characters. Very few methods have bugs (e.g. `trim` can trim individual bytes from UTF-8 characters that would be whitespace in Latin-1).

I wasn't aware of that docs statement, I will update it.

**Comment by phst** [comment]
Thanks!
What about the runfiles manifest? It can't be UTF-16 (that would be the appropriate OS encoding, but not ASCII-compatible), so is it UTF-8 even on Windows?

**Comment by phst** [comment]
> Starlark and other Bazel files (e.g. .bazelrc, manifests, ...) are assumed (but not enforced) to be encoded in UTF-8

Stardoc assumes Latin-1 for docstrings, though. Encoding a Starlark file in UTF-8 will result in double-encoding, cf. https://github.com/phst/rules_elisp/blob/master/docs/generate.py#L236-L239

**Comment by phst** [comment]
If Starlark files are now asssumed to be UTF-8, then I guess for Stardoc https://github.com/bazelbuild/bazel/blob/8.0.0/src/main/java/com/google/devtools/build/lib/starlarkdocextract/RuleInfoExtractor.java#L65 and similar occurrences (basically wherever a string proto field in the Stardoc proto is set) need to be fixed

**Comment by fmeum** [comment]
> Thanks! What about the runfiles manifest on Windows? It can't be UTF-16 (that would be the appropriate OS encoding, but not ASCII-compatible), so is it UTF-8 even on Windows?

Yes, all output files produced by Bazel should use UTF-8 and `\n` line endings on all platforms, including Windows.

> If Starlark files are now asssumed to be UTF-8, then I guess for Stardoc https://github.com/bazelbuild/bazel/blob/8.0.0/src/main/java/com/google/devtools/build/lib/starlarkdocextract/RuleInfoExtractor.java#L65 and similar occurrences (basically wherever a string proto field in the Stardoc proto is set) need to be fixed

Thanks for pointing that out, I sent https://github.com/bazelbuild/bazel/pull/24935 to fix this.

**Comment by phst** [comment]
> I wasn't aware of that docs statement, I will update it.

Here's another doc that I guess is outdated now: https://bazel.build/concepts/labels

- Repository names: No documentation, but I'd assume that even with this change repository names (both canonical and apparent) are ASCII-only, with some more restrictions (no newlines, spaces, slashes, ...)
- Package names: Since package names correspond to directory names, these can now also contain non-ASCII characters?
- Target names: Definitely can contain non-ASCII characters now.

**Comment by phst** [comment]
> Yes, all output files produced by Bazel should use UTF-8 and \n line endings on all platforms, including Windows.

OK, then the runfiles libraries also need to be adapted.
- Go: Probably no change needed, Go uses WTF-8 for filenames on Windows
- Python: Probably only need to make sure that files are opened with the correct encoding (https://github.com/bazelbuild/rules_python/pull/2568)
- C++: A lot more changes required. The use of narrow strings throughout makes manifest files and directories with non-ASCII characters nonportable/impossible. We need at least overloads for Create and Rlocation with std::wstring on Windows.
- Others: ???

**Comment by fmeum** [comment]
Thanks for sending the fix for Python!

> C++: A lot more changes required. The use of narrow strings throughout makes manifest files and directories with non-ASCII characters nonportable/impossible. We need at least overloads for Create and Rlocation with std::wstring on Windows.

Microsoft now [recommends](https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page) using the `-A` variety of functions with the UTF-8 code page (forced via an app manifest) instead of the wide string functions for new software. Existing software probably already has its own conversion functions to and from UTF-8, so I would personally not complicate the API for everyone by adding more overloads.

**Comment by phst** [comment]
> > Thanks! What about the runfiles manifest on Windows? It can't be UTF-16 (that would be the appropriate OS encoding, but not ASCII-compatible), so is it UTF-8 even on Windows?
> 
> Yes, all output files produced by Bazel should use UTF-8 and `\n` line endings on all platforms, including Windows.

It would also be great to put this statement into the official documentation.

**Comment by phst** [comment]
For reference, https://github.com/bazelbuild/bazel/commit/8a469b3364907e9b925e6531b1094200f8840c76 has updated https://bazel.build/concepts/build-files to reflect the current state regarding the encoding of BUILD files.

**Comment by sfc-gh-bhannel** [comment]
Setting the LANG environment variable before the build (as suggested [here](https://github.com/bazelbuild/bazel/issues/8108#issuecomment-579240732)) allowed me to build a target with non-ASCII files.
